src/styles.scss

```diff
@@ -6,6 +6,15 @@ body {
   color: var(--text-color-primary);
 }
 
+canvas {
+  // following props improve blurriness at certain devicePixelRatios.
+  // AFAIK it doesn't affect export (in fact, export seems sharp either way).
+
+  image-rendering: pixelated; // chromium
```

Comment:
This is likely the reason why the downsampling algorithm used is so bad.

Reply:
Nope, tried it without it. Same result.